tests/subprocesstest.py

```diff
@@ -0,0 +1,125 @@
+
+    Usage on a class::
+
+        @run_in_subprocess
+        class PatchTests(object):
+            # will be run in new interpreter
+            def test_patch_before_import(self):
+                patch()
+                import module
+
+            # will be run in new interpreter as well
+            def test_patch_after_import(self):
+                import module
+                patch()
+
+
+    Usage on a test method::
+
+        class OtherTests(object):
+            @run_in_subprocess
+            def test_case(self):
+                pass
+
+
+    :param obj: method or class to run in a separate python interpreter.
+    :return:
+    """
+    setattr(obj, SUBPROC_TEST_ATTR, True)
+    return obj
+
+
+class SubprocessTestCase(unittest.TestCase):
+    def _full_method_name(self):
+        test = getattr(self, self._testMethodName)
+        modpath = test.__module__
+        clsname = self.__class__.__name__
+        testname = test.__name__
+        testcase_name = '{}.{}.{}'.format(modpath, clsname, testname)
+        return testcase_name
+
+    def _run_test_in_subprocess(self, result):
+        full_testcase_name = self._full_method_name()
+
+        sp_test_cmd = ['python', '-m', 'unittest', full_testcase_name]
+        sp = subprocess.Popen(
+            sp_test_cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        _, stderr = sp.communicate()
+
+        if sp.returncode:
+            try:
+                cmdf = ' '.join(sp_test_cmd)
+                raise Exception('Subprocess Test "{}" Failed'.format(cmdf))
+            except Exception:
+                exc_info = sys.exc_info()
+            sys.stderr.write(stderr)
+            result.addFailure(self, exc_info)
+        else:
+            result.addSuccess(self)
+
+    def _in_subprocess(self):
+        """Determines if the test is being run in a subprocess.
+
+        This is done by checking the system arguments and seeing if the full
+        module method name is contained in any of the arguments. This method
+        assumes that the test case is being run in a subprocess if invoked with
+        a test command specifying only this test case.
+
+        For example the command:
+        $ python -m unittest tests.contrib.gevent.test_patch.TestGeventPatch.test_patch_before_import
+
+        will have _in_subprocess return True for the test_patch_before_import
+        test case for gevent.
+
+        :param test: the test case being run
+        :return: whether the test is being run individually (with the assumption
+                 that this is in a new subprocess)
+        """
+        full_testcase_name = self._full_method_name()
+        for arg in sys.argv:
+            if full_testcase_name in arg:
+                return True
+        return False
+
+    def _is_subprocess_test(self):
+        if hasattr(self, SUBPROC_TEST_ATTR):
+            return True
+
+        if hasattr(self, '_testMethodName'):
```

Comment:
Is there a case where this won't exist?

We assume it exists above in `self._full_method_name()`

Reply:
It is a private member.. but it's probably safe to rely on. I believe both pytest and nosetest depend on it as well.